uw-research-computing/singularity-htc.md

```diff
@@ -0,0 +1,167 @@
+
+The final `image.def` looks like:
+
+```
+Bootstrap: docker
+From: rocker/geospatial:4.2.2
+
+%post
+    apt-get update -y
+    apt-get install -y \
+            build-essential \
+            cmake \
+            g++ \
+            r-base-dev
+
+    R -e "install.packages('cowsay', dependencies=TRUE, repos='http://cran.rstudio.com/')"
+```
+
+## 4. Build your Apptainer Container
+
+Once your build spec is ready, you can "build" the container. The process of building an apptainer image will convert our `.def` file to a `.sif` Apptainer/Singularity image that we will use for future job submissions. To build our container, run this command:
+
+```
+$ apptainer build my-container.sif image.def
+```
+
+If you list the contents of your scratch directory by running `ls`, you should now see your executable Apptainer/Singularity file `my-container.sif`. 
+
+Once the image is built, it is important to test it to make sure you have all software, packages, and libraries installed correctly. 
+
+## 5. Testing Your Container
+
+Just like it is important to test your codes and jobs at a small scale, you should make sure that your container is working correctly. To do this, we will start an "interactive Apptainer session" with the Apptainer/Singularity "shell" mode. The recommended command line, similar to how containers are started for jobs, is:
+
+```
+THIS CODE NEEDS TESTING (LIKELY REMOVE /SRV, ETC)
+apptainer shell my-container.sif \
+            --home $PWD:/srv \
+            --pwd /srv \
+           #--bind /cvmfs \
+            --scratch /var/tmp \
+            --scratch /tmp \
+            --contain --ipc --pid \
+``` 
+
+This will give you an interactive shell using your container, with your current working directory mounted under /srv. Your prompt should change to: 
+
+```
+Apptainer my-container.sif:~>
+```
+
+You can explore the container and test your code in this mode. Once you are down exploring, exit the container by running `exit` or with `CTRL+D`.
+ 
+
+# 6. Exit your HTCondor Interactive Job
+
+Now that we have prepared our Apptainer image and tested it, we are ready to exit our interactive session. When we do, our build file will return to our `/home` directory on the submit server.  To exit, run:
+
+```
+[alice@build]$ exit
+```
+
+## 7. Use an Apptainer Container in HTC Jobs
+
+HTCondor will automatically fetch and  prepare your Apptainer environment for your high-throughput jobs as long as we provide HTCondor with the `.sif` file information in your job's submit file. Specifically, we need to provide HTCondor with the name of the `container_image` and to tell HTCondor where to find the `.sif` file to transfer to the job using `transfer_input_files` like so: 
+
+```
+# HTC Submit File
+
+# Provide HTCondor with the name of your .sif file and universe information
+container_image = my-container.sif
+universe = container
```

Comment:
Write a comment next to this that it is optional

Reply:
`universe = container` is optional?